File: projects/mcp-8004-enterprise/agents_implementation/vault-signer.js
/**
 * vault-signer.js
 *
 * HSM / Vault signing shim for oracle bridges.
 *
 * Replaces `new ethers.Wallet(privateKey, provider)` with a factory that
 * supports two backends:
 *
 *   "local"  — wraps ethers.Wallet (dev / CI). Same behaviour as before;
 *              private key is read from config.privateKey.
 *
 *   "vault"  — VaultSigner: extends ethers.AbstractSigner, delegates all
 *              signing operations to an HTTP signing endpoint (HashiCorp Vault
 *              Transit, AWS KMS, Azure Key Vault, or any compatible REST API).
 *              The private key never leaves the vault process.
 *
 * Usage in bridges:
 *
 *   import { createSigner } from './vault-signer.js';
 *
 *   const signer = await createSigner({
 *     type:       process.env.SIGNER_TYPE ?? 'local',   // 'local' | 'vault'
 *     privateKey: process.env.ORACLE_PRIVATE_KEY,       // used when type=local
 *     vaultUrl:   process.env.VAULT_URL,                // used when type=vault
 *     address:    process.env.ORACLE_ADDRESS,           // used when type=vault
 *   }, provider);
 *
 * The returned object implements the full ethers.Signer interface, so it can
 * be passed to `new ethers.Contract(address, abi, signer)` unchanged.
 */

import { ethers } from 'ethers';

// ── VaultSigner ───────────────────────────────────────────────────────────────

/**
 * Signer backed by an HTTP signing endpoint.
 *
 * Expected endpoint contract (POST ${vaultUrl}/sign):
 *   Request:  { "hash": "0x..." }           — hex-encoded 32-byte hash
 *   Response: { "signature": "0x..." }      — 65-byte ECDSA signature (r, s, v)
 *
 * This matches the HashiCorp Vault Transit sign API response shape after
 * adapting the base64-encoded signature. Production deployments should add
 * mTLS client certificates and vault token authentication headers.
 */
class VaultSigner extends ethers.AbstractSigner {
  #vaultUrl;
  #address;

  constructor(vaultUrl, address, provider) {
    super(provider);
    this.#vaultUrl = vaultUrl;
    this.#address  = address;
  }

  async getAddress() {
    return this.#address;
  }

  connect(provider) {
    return new VaultSigner(this.#vaultUrl, this.#address, provider);
  }

  async signTransaction(tx) {
    const populated = await this.populateTransaction(tx);
    const unsignedTx = ethers.Transaction.from(populated);
    const hash = ethers.keccak256(unsignedTx.unsignedSerialized);
    const signature = await this.#remoteSign(hash);
    unsignedTx.signature = signature;
    return unsignedTx.serialized;
  }

  async signMessage(message) {
    const hash = typeof message === 'string'
      ? ethers.hashMessage(message)
      : ethers.hashMessage(ethers.toUtf8String(message));
    return this.#remoteSign(hash);
  }

  async signTypedData(domain, types, value) {
    const hash = ethers.TypedDataEncoder.hash(domain, types, value);
    return this.#remoteSign(hash);
  }

  async #remoteSign(hash) {
    const res = await fetch(`${this.#vaultUrl}/sign`, {
      method:  'POST',
      headers: { 'Content-Type': 'application/json' },
      body:    JSON.stringify({ hash }),
    });
    if (!res.ok) {
      const text = await res.text().catch(() => '');
      throw new Error(`VaultSigner: signing endpoint returned ${res.status}: ${text}`);
    }
    const json = await res.json();
    if (!json.signature) throw new Error('VaultSigner: response missing "signature" field');
    return json.signature;
  }
}

// ── Factory ───────────────────────────────────────────────────────────────────

/**
 * @param {object} config
 * @param {string} config.type        'local' | 'vault'
 * @param {string} [config.privateKey] Required when type='local'
 * @param {string} [config.vaultUrl]   Required when type='vault' (e.g. 'http://vault:8200/v1/transit/keys/oracle')
 * @param {string} [config.address]    Required when type='vault' — the oracle wallet address
 * @param {ethers.Provider} provider
 * @returns {ethers.Signer}
 */
export function createSigner(config, provider) {
  const type = config.type ?? 'local';

  if (type === 'local') {
    if (!config.privateKey) throw new Error('vault-signer: config.privateKey required for type=local');
    return new ethers.Wallet(config.privateKey, provider);
  }

  if (type === 'vault') {
    if (!config.vaultUrl) throw new Error('vault-signer: config.vaultUrl required for type=vault');
    if (!config.address)  throw new Error('vault-signer: config.address required for type=vault');
    return new VaultSigner(config.vaultUrl, config.address, provider);
  }

  throw new Error(`vault-signer: unknown type "${type}". Use 'local' or 'vault'.`);
}

File: examples/mcp-8004-enterprise/agents_implementation_py/shared/vault_signer.py
"""
vault_signer.py — HSM / Vault signing shim for oracle bridges.
Port of vault-signer.js using web3.py Account abstraction.

Two backends:
  "local"  — wraps eth_account.Account (dev / CI)
  "vault"  — VaultSigner: delegates signing to an HTTP endpoint
             (HashiCorp Vault Transit, AWS KMS, Azure Key Vault, etc.)
             The private key never leaves the vault process.

Usage:
    from shared.vault_signer import create_signer, BaseSigner
    signer = create_signer(
        signer_type=os.getenv("SIGNER_TYPE", "local"),
        private_key=os.getenv("ORACLE_PRIVATE_KEY"),
        vault_url=os.getenv("VAULT_URL"),
        vault_address=os.getenv("ORACLE_ADDRESS"),
    )
    signed_tx = signer.sign_transaction(tx_dict)
    address   = signer.address
"""

from __future__ import annotations

import abc
from typing import Optional

import httpx
from eth_account import Account
from eth_account.datastructures import SignedTransaction


class BaseSigner(abc.ABC):
    """Minimal signer interface used by all bridges."""

    @property
    @abc.abstractmethod
    def address(self) -> str: ...

    @abc.abstractmethod
    def sign_transaction(self, tx: dict) -> SignedTransaction: ...

    @abc.abstractmethod
    def sign_message(self, message: bytes) -> bytes: ...


# ── Local signer (dev / CI) ───────────────────────────────────────────────────

class LocalSigner(BaseSigner):
    """Wraps eth_account.Account; private key held in process memory."""

    def __init__(self, private_key: str) -> None:
        self._account = Account.from_key(private_key)

    @property
    def address(self) -> str:
        return self._account.address

    def sign_transaction(self, tx: dict) -> SignedTransaction:
        return self._account.sign_transaction(tx)

    def sign_message(self, message: bytes) -> bytes:
        return self._account.sign_message(message).signature


# ── Vault signer (production) ─────────────────────────────────────────────────

class VaultSigner(BaseSigner):
    """
    Signer backed by an HTTP signing endpoint.

    Expected endpoint contract (POST ${vault_url}/sign):
      Request:  {"hash": "0x..."}           — hex-encoded 32-byte hash
      Response: {"signature": "0x..."}      — 65-byte ECDSA signature (r, s, v)
    """

    def __init__(self, vault_url: str, address: str) -> None:
        self._vault_url = vault_url.rstrip("/")
        self._address = address

    @property
    def address(self) -> str:
        return self._address

    def _remote_sign(self, hash_hex: str) -> str:
        resp = httpx.post(
            f"{self._vault_url}/sign",
            json={"hash": hash_hex},
            timeout=30,
        )
        resp.raise_for_status()
        data = resp.json()
        if "signature" not in data:
            raise ValueError("VaultSigner: response missing 'signature' field")
        return data["signature"]

    def sign_transaction(self, tx: dict) -> SignedTransaction:
        raise NotImplementedError(
            "VaultSigner.sign_transaction: full RLP reconstruction not yet implemented. "
            "Use web3.py middleware instead for production vault signing."
        )

    def sign_message(self, message: bytes) -> bytes:
        from eth_account.messages import encode_defunct
        from eth_utils import keccak

        signable = encode_defunct(primitive=message)
        msg_hash = "0x" + keccak(signable.body).hex()
        sig_hex = self._remote_sign(msg_hash)
        return bytes.fromhex(sig_hex.removeprefix("0x"))


# ── Factory ───────────────────────────────────────────────────────────────────

def create_signer(
    signer_type: str = "local",
    private_key: Optional[str] = None,
    vault_url: Optional[str] = None,
    vault_address: Optional[str] = None,
) -> BaseSigner:
    """
    Factory: returns the appropriate BaseSigner implementation.

    Args:
        signer_type:    "local" or "vault"
        private_key:    Required when signer_type="local"
        vault_url:      Required when signer_type="vault"
        vault_address:  Required when signer_type="vault"
    """
    if signer_type == "local":
        if not private_key:
            raise ValueError("create_signer: private_key required for signer_type='local'")
        return LocalSigner(private_key)

    if signer_type == "vault":
        if not vault_url:
            raise ValueError("create_signer: vault_url required for signer_type='vault'")
        if not vault_address:
            raise ValueError("create_signer: vault_address required for signer_type='vault'")
        return VaultSigner(vault_url, vault_address)

    raise ValueError(f"create_signer: unknown signer_type '{signer_type}'. Use 'local' or 'vault'.")
